src/v/cluster/tm_stm.h

```diff
@@ -42,6 +43,64 @@ namespace cluster {
 using use_tx_version_with_last_pid_bool
   = ss::bool_class<struct use_tx_version_with_last_pid_tag>;
 
+class tm_stm;
+
+class txlock_unit {
+    tm_stm* _stm;
+    ss::lw_shared_ptr<ssx::semaphore_units> _units;
+    kafka::transactional_id _id;
+    std::string_view _name;
+
+public:
+    txlock_unit(
+      tm_stm* stm,
+      ss::lw_shared_ptr<ssx::semaphore_units> units,
+      kafka::transactional_id id,
+      std::string_view name) noexcept
+      : _stm(stm)
+      , _units(units)
+      , _id(id)
+      , _name(name) {
+        vlog(txlog.trace, "got_lock name:{}, tx_id:{}", _name, _id);
+    }
+
+    txlock_unit(txlock_unit&& token) noexcept
+      : _stm(token._stm)
+      , _units(token._units)
+      , _id(token._id)
+      , _name(token._name) {
+        token._stm = nullptr;
+        token._units = nullptr;
+    }
+    txlock_unit(const txlock_unit&) = delete;
+    txlock_unit(const txlock_unit&&) = delete;
+    txlock_unit() = delete;
+
+    bool return_all() {
+        if (_stm) {
+            _units->return_all();
+            _stm = nullptr;
+            _units = nullptr;
+            return true;
+        }
+        return false;
+    }
+
+    txlock_unit& operator=(txlock_unit&& other) noexcept {
+        if (this != &other) {
+            _stm = other._stm;
+            _units = other._units;
+            _id = other._id;
+            _name = other._name;
+            other._stm = nullptr;
+            other._units = nullptr;
+        }
+        return *this;
+    }
+
+    ~txlock_unit() noexcept;
```

Comment:
why is desctructor not defined ?

Reply:
destructor is using `tm_stm` so it should be defined after it